org.openscenegraph.android/jni/screenview.cpp

```diff
@@ -0,0 +1,389 @@
+    }
+
+    osg::Image* image = _imageBuffer[_currentImageIndex].get();
+
+    osg::Timer_t tick_start = osg::Timer::instance()->tick();
+
+#if 1
+    image->readPixels(0,0,_width,_height,_pixelFormat,_type);
+
+#endif
+
+    osg::Timer_t tick_afterReadPixels = osg::Timer::instance()->tick();
+
+    updateTimings(tick_start, tick_afterReadPixels, tick_afterReadPixels, image->getTotalSizeInBytes());
+
+    _image_available = true;
+    _currentImageToTransfer = reinterpret_cast<osg::Image*>( image->clone ( osg::CopyOp::DEEP_COPY_ALL ) );
+
+    if (!_fileName.empty())
+    {
+        // osgDB::writeImageFile(*image, _fileName);
+    }
+
+    _currentImageIndex = nextImageIndex;
+    _currentPboIndex = nextPboIndex;
+}
+
+void WindowCaptureCallback::ContextData::singlePBO(osg::GLExtensions* ext)
+{
+    // std::cout<<"singelPBO(  "<<_fileName<<" image "<<_currentImageIndex<<" "<<_currentPboIndex<<std::endl;
+
+    unsigned int nextImageIndex = (_currentImageIndex+1)%_imageBuffer.size();
+
+    int width=0, height=0;
+    getSize(_gc, width, height);
+    if (width!=_width || _height!=height)
+    {
+        //std::cout<<"   Window resized "<<width<<", "<<height<<std::endl;
+        _width = width;
+        _height = height;
+    }
+
+    GLuint& pbo = _pboBuffer[0];
+
+    osg::Image* image = _imageBuffer[_currentImageIndex].get();
+    if (image->s() != _width ||
+        image->t() != _height)
+    {
+#ifdef DEBUG
+        osg::notify(osg::NOTICE)<<"Allocating image "<<std::endl;
+#endif
+        image->allocateImage(_width, _height, 1, _pixelFormat, _type);
+
+        if (pbo!=0)
+        {
+#ifdef DEBUG
+            osg::notify(osg::NOTICE)<<"deleting pbo "<<pbo<<std::endl;
+#endif
+            ext->glDeleteBuffers (1, &pbo);
+            pbo = 0;
+        }
+    }
+
+
+    if (pbo==0)
+    {
+        ext->glGenBuffers(1, &pbo);
+        ext->glBindBuffer(GL_PIXEL_PACK_BUFFER_ARB, pbo);
+        ext->glBufferData(GL_PIXEL_PACK_BUFFER_ARB, image->getTotalSizeInBytes(), 0, GL_STREAM_READ);
+#ifdef DEBUG
+        osg::notify(osg::NOTICE)<<"Generating pbo "<<pbo<<std::endl;
+#endif
+    }
+    else
+    {
+        ext->glBindBuffer(GL_PIXEL_PACK_BUFFER_ARB, pbo);
+    }
+
+    osg::Timer_t tick_start = osg::Timer::instance()->tick();
+
+#if 1
+    glReadPixels(0, 0, _width, _height, _pixelFormat, _type, 0);
+#endif
+
+    osg::Timer_t tick_afterReadPixels = osg::Timer::instance()->tick();
+
+    GLubyte* src = (GLubyte*)ext->glMapBuffer(GL_PIXEL_PACK_BUFFER_ARB,
+                                              GL_READ_ONLY_ARB);
+    if(src)
+    {
+        memcpy(image->data(), src, image->getTotalSizeInBytes());
+
+        ext->glUnmapBuffer(GL_PIXEL_PACK_BUFFER_ARB);
+    }
+
+    ext->glBindBuffer(GL_PIXEL_PACK_BUFFER_ARB, 0);
+
+    osg::Timer_t tick_afterMemCpy = osg::Timer::instance()->tick();
+
+    updateTimings(tick_start, tick_afterReadPixels, tick_afterMemCpy, image->getTotalSizeInBytes());
+
+    if (!_fileName.empty())
```

Comment:
I just copied the plain example for offscreen rendering from the web, which worked on the desktop and which I wrapped as-is. I can have a look on it again, but it may also be helpful if you could point me to specifics in the function that are 'dead' by now.